File: src/mini/passes/syntax.nim
#:_______________________________________________________________________
#  mini.nim  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU GPLv3 or later  :
#:_______________________________________________________________________
# @deps std
from std/strformat import `&`
from std/strutils  import `%`, join
# @deps slate
import slate except fail, Pos
# @deps mini.nim
import ../types/base
import ../types/parser
import ../types/tokenizer
import ../types/ast
import ../rules as mini
import ../errors


#_______________________________________
# @section Parser: Whitespace & Newlines
#_____________________________
func newline *(P :var Parser) :void=
  # TODO: Shouldn't ignore empty newlines. They matter for meaningful indentation
  # P.skip wht_newline
  P.move(1)
#___________________
func indentation *(P :var Parser; list :varargs[TokenID]) :void=
  # TODO: Meaningful indentation
  P.skip wht_space


#_______________________________________
# @section Parse: Expressions
#_____________________________
func literal *(P :var Parser) :ast.Expression=
  result = ast.Expression(kind: Literal)
  P.expect TokenID.b_number
  result.lit_value = P.token.loc.From(P.src)
#_____________________________
func expression *(P :var Parser) :ast.Expression=
  result = P.literal()


#_______________________________________
# @section Parse: Statements
#_____________________________
const StatementIDs * = {TokenID.kw_return, TokenID.kw_var}
#___________________
func statement_return *(P :var Parser) :ast.Statement=
  P.expect TokenID.kw_return
  P.move(1)
  P.indentation()
  result = ast.Statement(kind: Return)
  result.ret_value = P.expression()
#___________________
func statement_variable *(P :var Parser) :ast.Statement=
  P.expect TokenID.kw_var
  P.move(1)
  P.indentation()
  result = ast.Statement(kind: Variable)
  # Name
  P.expect TokenID.b_ident
  result.var_name = P.token.loc.From(P.src)
  P.move(1)
  P.indentation()
  # Type
  # FIX: Parse Type
  # Value
  P.expect TokenID.sp_equal, TokenID.sp_semicolon
  P.move(1)
  P.indentation()
  result.var_value = P.expression()
#___________________
func statement *(P :var Parser) :ast.Statement=
  P.expect TokenID.kw_return, TokenID.kw_var
  result = case P.token.id
  of kw_return : P.statement_return()
  of kw_var    : P.statement_variable()
  else         : UnknownStatementTokenError.fail &"Parsing token `{P.token}` as a statement is not implemented."; Statement()


#_______________________________________
# @section Parse: proc
#_____________________________
func Proc_args *(P :var Parser) :ast.Proc_Args=
  result = @[]
  P.expect TokenID.sp_paren_L
  P.move(1)
  P.indentation()
  # FIX: Add arguments
  P.expect TokenID.sp_paren_R
  P.move(1)
  P.indentation()
#_____________________________
func Proc_body *(P :var Parser; proc_depth :slate.Depth) :Proc_Body=
  result = @[]
  P.expect TokenID.sp_equal
  P.move(1)
  P.indentation()
  # Might have a newline at the start of the body
  if P.token.id == TokenID.wht_newline:
    P.move(1)
    P.indentation()
  # Parse all statements
  while proc_depth in P.token.depth:
    P.indentation()
    result.add P.statement()
    P.move(1)
    P.indentation()
    if P.last: break
#_____________________________
func Proc_retT *(P :var Parser) :ast.Type=
  result = ast.Type()
  P.expect TokenID.sp_colon
  P.move(1)
  P.indentation()
  P.expect TokenID.b_ident
  result.name = P.token.loc.From(P.src)
  P.move(1)
  P.indentation()
#_____________________________
func Proc *(P :var Parser) :void=
  var res = ast.Node(kind: Proc)
  # Skip Keyword
  P.expect TokenID.kw_proc
  let proc_depth = P.token.depth
  P.move(1)
  P.indentation()
  # Get the name
  P.expect TokenID.b_ident
  res.name = P.token.loc.From(P.src)
  P.move(1)
  P.indentation()
  # Make public when marked with *
  if P.token.id == op_star:
    res.public = true
    P.move(1)
    P.indentation()
  # Get the Args
  P.expect TokenID.sp_paren_L
  res.proc_args = P.Proc_args()
  # Get the Return Type
  P.expect TokenID.sp_colon
  res.proc_retT = P.Proc_retT()
  # Get the body
  P.expect TokenID.sp_equal, TokenID.sp_semicolon
  if P.token.id == TokenID.sp_equal:
    res.proc_body = P.Proc_body(proc_depth)
  elif P.token.id == TokenID.sp_semicolon:
    P.move(1)
    P.indentation()
  # Add the proc node to the AST
  P.ast.nodes.add res

#_______________________________________
# @section Parser: var
#_____________________________
func Var_type *(P :var Parser) :ast.Var_type=
  result = ast.Var_type()
  result.name = "int"
#___________________
func Var_value *(P :var Parser) :ast.Var_value=
  P.expect TokenID.sp_equal
  P.move(1)
  P.indentation()
  result = P.literal()
#___________________
func variable *(P :var Parser) :void=
  var res = ast.Node(kind: Var)
  # Skip Keyword
  P.expect TokenID.kw_var
  P.move(1)
  P.indentation()
  # Get the name
  P.expect TokenID.b_ident
  res.name = P.token.loc.From(P.src)
  P.move(1)
  P.indentation()
  # Make public when marked with *
  if P.token.id == op_star:
    res.public = true
    P.move(1)
    P.indentation()
  # TODO: Get the Value Type
  res.var_type = P.Var_type()
  # Get the Value
  P.expect TokenID.sp_equal, TokenID.sp_semicolon
  if P.token.id == TokenID.sp_equal:
    res.var_value = P.Var_value()
    P.move(1)
  elif P.token.id == TokenID.sp_semicolon:
    P.move(1)
    P.indentation()
  # Expect a newline at the end of the statement
  P.expect TokenID.wht_newline
  # P.move(1)  # @note: Do not move to the next token, the main while loop will do that
  # Add the var node to the AST
  P.ast.nodes.add res


func pass *(P :var Parser) :void=
  ## @descr
  ## Entry point of the Syntax Parser pass.
  ## Will process the TokenList of the Parser into the resulting AST.
  while P.pos < P.buf.len.Sz:
    case P.token.id
    of kw_proc     : P.Proc()
    of kw_var      : P.variable()
    of wht_newline : P.newline()
    else           : P.error_toplevel()
    P.pos.inc

File: src/mini/passes/syntax.c
//:____________________________________________________________
//  mini.nim  |  Copyright (C) Ivan Mar (sOkam!)  |  MPL-2.0  :
//:____________________________________________________________
#include "../list.h"
#include "../tokenizer.h"
#include "../parser/errors.h"
#include "../passes.h"


static void mini_parser_add (
  mini_Parser* const     P,
  mini_Node const* const node
) {
  mini_list_grow((mini_List*)&P->ast.nodes, 1, sizeof(*P->ast.nodes.ptr));
  P->ast.nodes.ptr[P->ast.nodes.len - 1] = *node;
}

static void mini_parser_syntax_statement_list_add (
  mini_statement_List* const  list,
  mini_Statement const* const statement
) {
  mini_list_grow((mini_List*)list, 1, sizeof(*list->ptr));
  list->ptr[list->len - 1] = *statement;
}


//______________________________________
// @section Parse Syntax: Expressions
//____________________________

static mini_Expression mini_parser_syntax_expression_literal (
  mini_Parser* const P
) {
  mini_parser_expect_any(P, 1, (mini_token_Id[]){ mini_token_b_number });
  mini_Expression result = (mini_Expression) /* clang-format off */ {
    .kind    = mini_expression_literal,
    .data    = (mini_expression_Data){.literal= (mini_Literal){.number= (mini_literal_Number){
      .value = P->buf.ptr[P->pos].loc,
    }}}
  };  // clang-format on
  P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_sp_semicolon) P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_newline) P->pos += 1;
  return result;
}

static mini_Expression mini_parser_syntax_expression (
  mini_Parser* const P
) {
  mini_parser_expect_any(P, 1, (mini_token_Id[]){ mini_token_b_number, mini_token_b_identifier });
  mini_Token const tk = P->buf.ptr[P->pos];
  switch (tk.id) {
    case mini_token_b_number : return mini_parser_syntax_expression_literal(P); break;
    // case mini_token_b_identifier : mini_parser_syntax_expression_identifier(P); break;
    default                  : mini_parser_error(P, "Unknown First Token for Expression: %02zu:%s", P->pos, mini_token_toString(tk.id)); break;
  }
}


//______________________________________
// @section Parse Syntax: statement
//____________________________

static mini_Var mini_parser_syntax_statement_var (
  mini_Parser* const P
) {
  mini_parser_expect(P, mini_token_kw_var);
  P->pos += 1;
  // Create the result
  mini_Var result = (mini_Var){ 0 };
  // Continue parsing the var
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  // Assign Name
  mini_parser_expect(P, mini_token_b_identifier);
  result.name = P->buf.ptr[P->pos].loc;
  P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  // Assign Visibility
  if (P->buf.ptr[P->pos].id == mini_token_sp_star) {
    result.visibility = mini_public;
    P->pos += 1;
  }
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  // Type
  mini_parser_expect(P, mini_token_sp_colon);
  P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  mini_parser_expect(P, mini_token_b_identifier);
  result.type = P->buf.ptr[P->pos].loc;
  P->pos += 1;
  if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
  // Value
  mini_parser_expect(P, mini_token_sp_equal);
  P->pos += 1;
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  result.value = mini_parser_syntax_expression(P);
  return result;
}

static mini_Return mini_parser_syntax_statement_return (
  mini_Parser* const P
) {
  mini_parser_expect(P, mini_token_kw_return);
  P->pos += 1;
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  mini_Return result = (mini_Return){ 0 };
  result.value       = mini_parser_syntax_expression(P);
  return result;
}

static mini_statement_List mini_parser_syntax_statement_list (
  mini_Parser* const P
) {
  mini_parser_expect_any(P, 2, (mini_token_Id[]){ mini_token_wht_space, mini_token_wht_newline });
  P->pos += 1;
  mini_statement_List result        = (mini_statement_List){ 0 };
  mini_token_Scope    depth_current = P->buf.ptr[P->pos].depth;
  while (P->pos < P->buf.len) {  // FIX: Stop when we exit the scope
    if (P->buf.ptr[P->pos].id == mini_token_wht_newline) {
      P->pos += 1;               // Skip empty lines
      continue;
    }
    // Stop if we changed scope
    if (P->buf.ptr[P->pos].depth.scope != depth_current.scope) break;
    depth_current = P->buf.ptr[P->pos].depth;
    // Skip whitespace at the start of the next line
    if (P->buf.ptr[P->pos].id == mini_token_wht_space) P->pos += 1;
    // Start parsing
    mini_parser_expect_any(P, 2, (mini_token_Id[]){ mini_token_kw_var, mini_token_kw_return });
    mini_Statement   statement = (mini_Statement){ 0 };
    mini_Token const tk        = P->buf.ptr[P->pos];
    switch (tk.id) {
      case mini_token_kw_var : {
        statement.kind = mini_statement_var;
        statement.data = (mini_statement_Data){ .var = mini_parser_syntax_statement_var(P) };
        break;
      }
      case mini_token_kw_return : {
        statement.kind = mini_statement_return;
        statement.data = (mini_statement_Data){ .Return = mini_parser_syntax_statement_return(P) };
        break;
      }
      default : mini_parser_error(P, "Unknown First Token for Statement: %02zu:%s", P->pos, mini_token_toString(tk.id)); break;
    }
    mini_parser_syntax_statement_list_add(&result, &statement);
  }
  return result;
}


//______________________________________
// @section Parse Syntax: var
//____________________________

static void mini_parser_syntax_var (
  mini_Parser* const P
) {
  mini_parser_expect(P, mini_token_kw_var);
  mini_Node result = (mini_Node){
    .kind = mini_node_var,
    .data = mini_parser_syntax_statement_var(P),
  };
  // Add Node to the AST
  mini_parser_add(P, &result);
  P->pos -= 1;
}


//______________________________________
// @section Parse Syntax: proc
//____________________________

static void mini_parser_syntax_proc (
  mini_Parser* const P
) {
  mini_parser_expect(P, mini_token_kw_proc);
  P->pos += 1;
  // Create the result
  mini_Node result = (mini_Node){
    .kind = mini_node_proc,
    .data = (mini_node_Data){ .proc = (mini_Proc){ 0 } },
  };
  // Continue parsing the proc
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  // Assign Name
  mini_parser_expect(P, mini_token_b_identifier);
  result.data.proc.name = P->buf.ptr[P->pos].loc;
  P->pos += 1;
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  // Assign Visibility
  if (P->buf.ptr[P->pos].id == mini_token_sp_star) {
    result.data.proc.visibility = mini_public;
    P->pos += 1;
  }
  // Proc Arguments
  mini_parser_expect(P, mini_token_parenthesis_left);
  P->pos += 1;
  // FIX: Parse the Arguments
  mini_parser_expect(P, mini_token_parenthesis_right);
  P->pos += 1;
  mini_parser_expect(P, mini_token_wht_space);
  P->pos += 1;
  // Return Type
  mini_parser_expect(P, mini_token_sp_colon);
  P->pos += 1;
  mini_parser_expect(P, mini_token_b_identifier);
  result.data.proc.return_type = P->buf.ptr[P->pos].loc;
  P->pos += 1;
  mini_parser_expect(P, mini_token_sp_equal);
  P->pos += 1;
  // Proc Body
  result.data.proc.body = mini_parser_syntax_statement_list(P);
  // Add Node to the AST
  mini_parser_add(P, &result);
  P->pos -= 1;
}


//______________________________________
// @section Parser.syntax: Entry Point
//____________________________

void mini_parser_syntax (
  mini_Parser* const P
) {
  while (P->pos < P->buf.len) {
    mini_Token const tk = P->buf.ptr[P->pos];
    switch (tk.id) {
      case mini_token_kw_proc     : mini_parser_syntax_proc(P); break;
      case mini_token_kw_var      : mini_parser_syntax_var(P); break;
      case mini_token_wht_space   :         /* fall-through */
      case mini_token_wht_newline : break;  // FIX: mini_parser_syntax_whitespace(P);
      default                     : mini_parser_error(P, "Unknown TopLevel Token: %02zu:%s", P->pos, mini_token_toString(tk.id)); break;
    }
    P->pos += 1;
  }
  P->pos = 0;
}

